Fix GRIN→LLVM codegen for programs with ADTs and higher-order functions

## Context

After PR #411 (issue #410: TagTable, Case dispatch, Store, and nullary constructors), the compiler could handle `putStrLn $ id "Hello"`. However, compiling programs with user-defined ADTs and higher-order functions (e.g., `map` over a custom `List`) failed at the GRIN→LLVM codegen stage.

## Bugs

### 1. Unique=0 in synthetic lambda binders (Core desugarer)
`desugarMatch` in `src/core/desugar.zig` created synthetic lambda binders with `unique = .{ .value = 0 }`. When a function had multiple synthetic parameters (e.g., `map_it f list`), both binders shared the same unique, causing GRIN translation to produce duplicate parameter names (`arg_0_3 arg_0_3` instead of `arg_0_3 arg_1_4`).

**Fix:** Add `UniqueSupply` to `DesugarCtx` and use `u_supply.fresh()` for each synthetic binder.

### 2. `translateLet` NonRec used `freshName("p")` instead of the binder's name
In `src/grin/translate.zig`, the `translateLet` NonRec case used `freshName("p")` for the bind pattern instead of `mapBinder(&pair.binder)`. This disconnected the let-bound variable name from the body's references to it.

**Fix:** Use `mapBinder(&pair.binder)` for the bind pattern.

### 3. `params` map keyed by base name instead of unique ID (GRIN→LLVM codegen)
In `src/backend/grin_to_llvm.zig`, the `params` map used `name.base` (e.g., `"arg"`) as its key. Variables with the same base but different unique IDs (e.g., `arg_20`, `arg_21`) would overwrite each other, causing "Var arg not found" errors.

**Fix:** Change `params` from `StringHashMap` to `AutoHashMap(u64, ...)` keyed by `name.unique.value`.

### 4. `translateExprToValue` did not handle `Bind` expressions
Left-associative ANF bind chains (produced by `wrapWithPendingBinds`) have their `lhs` as another `Bind`. The `translateExprToValue` `else` branch called `translateExpr(Bind(...))` but returned `null`, losing the final value and preventing outer variables from being bound.

**Fix:** Add a `.Bind` arm to `translateExprToValue` that recursively evaluates the chain, binding each inner variable into `params`, and returns the final rhs value.

### 5. Higher-order calls emitted as direct calls to undefined functions
When `translateApp` encountered a callee name in `params` (a function pointer, e.g., `f_5` holding `identity`), it fell through to `LLVMGetNamedFunction("f")` → not found → created a forward declaration. This caused linker errors (`undefined reference to 'f'`).

**Fix:** Check `params.get(name.unique.value)` first in `translateApp`/`translateAppToValue`; if found, use the pointer for an indirect call via `LLVMBuildCall2`.

## Deliverable
- All five bugs fixed
- `hello.hs` with custom `List` ADT, `map_it`, `head_it`, and `identity` compiles and runs correctly
- 692/692 tests pass
- Golden files updated for `ghc_016` and `ghc_017` (synthetic binder names now include unique IDs)

## References
- `src/core/desugar.zig` — UniqueSupply fix
- `src/grin/translate.zig` — translateLet binder fix  
- `src/backend/grin_to_llvm.zig` — params map, Bind handling, indirect calls
- #410 (previous TagTable/Case work)

--- a/src/backend/grin_to_llvm.zig
+++ b/src/backend/grin_to_llvm.zig
@@ -328,8 +328,10 @@ pub const GrinTranslator = struct {
     module: llvm.Module,
     builder: llvm.Builder,
     allocator: std.mem.Allocator,
-    /// Maps GRIN variable names to their current LLVM SSA values.
-    params: std.StringHashMap(llvm.Value),
+    /// Maps GRIN variable unique IDs to their current LLVM SSA values.
+    /// Keyed by `Name.unique.value` so that variables with the same base name
+    /// but different uniques (e.g. `arg_20`, `arg_21`) are kept distinct.
+    params: std.AutoHashMap(u64, llvm.Value),
     /// Constructor tag discriminants, built before translation.
     tag_table: TagTable,
     /// The LLVM function currently being translated (needed for
@@ -344,7 +346,7 @@ pub const GrinTranslator = struct {
             .module = llvm.createModule("haskell", ctx),
             .builder = llvm.createBuilder(ctx),
             .allocator = allocator,
-            .params = std.StringHashMap(llvm.Value).init(allocator),
+            .params = std.AutoHashMap(u64, llvm.Value).init(allocator),
             .tag_table = TagTable.init(),
             .current_func = null,
         };
@@ -403,13 +405,13 @@ pub const GrinTranslator = struct {
 
         // Clear previous function's parameter mapping and set up current one.
         self.params.deinit();
-        self.params = std.StringHashMap(llvm.Value).init(self.allocator);
+        self.params = std.AutoHashMap(u64, llvm.Value).init(self.allocator);
 
         // Store each parameter as a named value in the params map.
         for (def.params, 0..) |param_name, i| {
             const param_val = c.LLVMGetParam(func, @intCast(i));
             if (param_val == null) return error.OutOfMemory;
-            try self.params.put(param_name.base, param_val);
+            try self.params.put(param_name.unique.value, param_val);
         }
 
         try self.translateExpr(def.body);
@@ -457,7 +459,7 @@ pub const GrinTranslator = struct {
                 // Bind the LHS result to a named variable.
                 const result = try self.translateExprToValue(lhs, v.base);
                 if (result) |val| {
-                    try self.params.put(v.base, val);
+                    try self.params.put(v.unique.value, val);
                 }
                 try self.translateExpr(rhs);
             },
@@ -471,7 +473,7 @@ pub const GrinTranslator = struct {
                 const struct_ty = nodeStructType(n_fields);
                 for (ctn.fields, 0..) |field_val, fi| {
                     if (field_val != .Var) continue;
-                    const field_name = field_val.Var.base;
+                    const field_var = field_val.Var;
                     // GEP: index 0 = tag, index fi+1 = field fi.
                     var indices = [_]llvm.Value{
                         c.LLVMConstInt(llvm.i32Type(), 0, 0),
@@ -483,10 +485,10 @@ pub const GrinTranslator = struct {
                         ptr_val,
                         &indices,
                         2,
-                        nullTerminate(field_name).ptr,
+                        nullTerminate(field_var.base).ptr,
                     );
-                    const loaded = c.LLVMBuildLoad2(self.builder, ptrType(), gep, nullTerminate(field_name).ptr);
-                    try self.params.put(field_name, loaded);
+                    const loaded = c.LLVMBuildLoad2(self.builder, ptrType(), gep, nullTerminate(field_var.base).ptr);
+                    try self.params.put(field_var.unique.value, loaded);
                 }
                 try self.translateExpr(rhs);
             },
@@ -496,6 +498,10 @@ pub const GrinTranslator = struct {
 
     /// Translate an expression and return the resulting LLVM value.
     /// Returns null for expressions that produce no SSA value (void calls).
+    ///
+    /// For `Bind(lhs, pat, rhs)` — which arises from left-associative ANF chains —
+    /// we evaluate the inner bind sequence imperatively (each step binding its
+    /// result into `params`) and then return the value of the final `rhs`.
     fn translateExprToValue(
         self: *GrinTranslator,
         expr: *const grin.Expr,
@@ -505,6 +511,40 @@ pub const GrinTranslator = struct {
             .App => |app| self.translateAppToValue(app.name, app.args, result_name),
             .Return => |v| @as(?llvm.Value, try self.translateValToLlvm(v)),
             .Store => |v| self.translateStoreToValue(v, result_name),
+            .Bind => |b| {
+                // Evaluate the lhs, bind result to pat, then return the value of rhs.
+                // This handles left-leaning ANF bind chains produced by wrapWithPendingBinds.
+                const inner_name: []const u8 = switch (b.pat) {
+                    .Var => |v| v.base,
+                    else => result_name,
+                };
+                const inner_result = try self.translateExprToValue(b.lhs, inner_name);
+                switch (b.pat) {
+                    .Var => |v| {
+                        if (inner_result) |val| try self.params.put(v.unique.value, val);
+                    },
+                    .Unit => {},
+                    .ConstTagNode => |ctn| {
+                        if (inner_result) |ptr_val| {
+                            const n_fields: u32 = @intCast(ctn.fields.len);
+                            const struct_ty = nodeStructType(n_fields);
+                            for (ctn.fields, 0..) |field_val, fi| {
+                                if (field_val != .Var) continue;
+                                const fv = field_val.Var;
+                                var indices = [_]llvm.Value{
+                                    c.LLVMConstInt(llvm.i32Type(), 0, 0),
+                                    c.LLVMConstInt(llvm.i32Type(), @intCast(fi + 1), 0),
+                                };
+                                const gep = c.LLVMBuildGEP2(self.builder, struct_ty, ptr_val, &indices, 2, nullTerminate(fv.base).ptr);
+                                const loaded = c.LLVMBuildLoad2(self.builder, ptrType(), gep, nullTerminate(fv.base).ptr);
+                                try self.params.put(fv.unique.value, loaded);
+                            }
+                        }
+                    },
+                    else => return error.UnimplementedPattern,
+                }
+                return self.translateExprToValue(b.rhs, result_name);
+            },
             else => {
                 try self.translateExpr(expr);
                 return null;
@@ -544,20 +584,24 @@ pub const GrinTranslator = struct {
             llvm_args[i] = try self.translateValToLlvm(val);
         }
 
-        var fn_name_buf: [128]u8 = undefined;
-        std.debug.assert(name.base.len < fn_name_buf.len);
-        @memcpy(fn_name_buf[0..name.base.len], name.base);
-        fn_name_buf[name.base.len] = 0;
-        const fn_name_z = fn_name_buf[0..name.base.len :0];
-        const func = blk: {
-            if (c.LLVMGetNamedFunction(self.module, fn_name_z)) |existing| {
-                break :blk existing;
-            }
-            var param_types: [8]llvm.Type = undefined;
-            for (0..arg_count) |i| {
-                param_types[i] = c.LLVMTypeOf(llvm_args[i]);
-            }
-            const fn_type = llvm.functionType(ptrType(), param_types[0..arg_count], false);
+        // Build a ptr-returning function type for indirect/forward calls.
+        var param_types: [8]llvm.Type = undefined;
+        for (0..arg_count) |i| param_types[i] = ptrType();
+        const fn_type = llvm.functionType(ptrType(), param_types[0..arg_count], false);
+
+        // Resolve the callee:
+        //   1. Local variable holding a function pointer (higher-order call) — use
+        //      an indirect call via the pointer stored in `params`.
+        //   2. Named LLVM function (direct call to a known def).
+        //   3. Forward-declare and call (for functions not yet translated).
+        const func: llvm.Value = blk: {
+            if (self.params.get(name.unique.value)) |fn_ptr| break :blk fn_ptr;
+            var fn_name_buf: [128]u8 = undefined;
+            std.debug.assert(name.base.len < fn_name_buf.len);
+            @memcpy(fn_name_buf[0..name.base.len], name.base);
+            fn_name_buf[name.base.len] = 0;
+            const fn_name_z = fn_name_buf[0..name.base.len :0];
+            if (c.LLVMGetNamedFunction(self.module, fn_name_z)) |existing| break :blk existing;
             break :blk llvm.addFunction(self.module, name.base, fn_type);
         };
 
@@ -570,7 +614,7 @@ pub const GrinTranslator = struct {
 
         return c.LLVMBuildCall2(
             self.builder,
-            llvm.getFunctionType(func),
+            fn_type,
             func,
             @ptrCast(&llvm_args),
             @intCast(arg_count),
@@ -595,25 +639,24 @@ pub const GrinTranslator = struct {
         for (args[0..arg_count], 0..) |val, i| {
             llvm_args[i] = try self.translateValToLlvm(val);
         }
-        var fn_name_buf: [128]u8 = undefined;
-        std.debug.assert(name.base.len < fn_name_buf.len);
-        @memcpy(fn_name_buf[0..name.base.len], name.base);
-        fn_name_buf[name.base.len] = 0;
-        const fn_name_z = fn_name_buf[0..name.base.len :0];
-        const func = blk: {
-            if (c.LLVMGetNamedFunction(self.module, fn_name_z)) |existing| {
-                break :blk existing;
-            }
-            var param_types: [8]llvm.Type = undefined;
-            for (0..arg_count) |i| {
-                param_types[i] = c.LLVMTypeOf(llvm_args[i]);
-            }
-            const fn_type = llvm.functionType(ptrType(), param_types[0..arg_count], false);
+
+        var param_types: [8]llvm.Type = undefined;
+        for (0..arg_count) |i| param_types[i] = ptrType();
+        const fn_type = llvm.functionType(ptrType(), param_types[0..arg_count], false);
+
+        const func: llvm.Value = blk: {
+            if (self.params.get(name.unique.value)) |fn_ptr| break :blk fn_ptr;
+            var fn_name_buf: [128]u8 = undefined;
+            std.debug.assert(name.base.len < fn_name_buf.len);
+            @memcpy(fn_name_buf[0..name.base.len], name.base);
+            fn_name_buf[name.base.len] = 0;
+            const fn_name_z = fn_name_buf[0..name.base.len :0];
+            if (c.LLVMGetNamedFunction(self.module, fn_name_z)) |existing| break :blk existing;
             break :blk llvm.addFunction(self.module, name.base, fn_type);
         };
         _ = c.LLVMBuildCall2(
             self.builder,
-            llvm.getFunctionType(func),
+            fn_type,
             func,
             @ptrCast(&llvm_args),
             @intCast(arg_count),
@@ -682,14 +725,28 @@ pub const GrinTranslator = struct {
             },
             .Unit => c.LLVMGetUndef(llvm.i32Type()),
             .Var => |name| blk: {
-                // 1. Check params (function parameters and let-bound variables).
-                if (self.params.get(name.base)) |v| break :blk v;
+                // 1. Check params (function parameters and let-bound variables),
+                //    keyed by unique ID to distinguish variables with the same base.
+                if (self.params.get(name.unique.value)) |v| break :blk v;
                 // 2. Nullary constructor: emit its tag discriminant as an i64.
                 //    Tracked in: https://github.com/adinapoli/rusholme/issues/410
                 if (self.tag_table.discriminantByName(name)) |disc| {
                     break :blk c.LLVMConstInt(llvm.i64Type(), @bitCast(disc), 0);
                 }
-                std.debug.print("UnsupportedGrinVal: Var {s} not found in params or tag table\n", .{name.base});
+                // 3. Top-level function reference: look up the LLVM function by
+                //    name and return its pointer (for higher-order use).
+                //    This handles passing named functions as arguments, e.g.:
+                //      map_it identity xs  =>  identity is a global function ptr.
+                var fn_name_buf: [128]u8 = undefined;
+                if (name.base.len < fn_name_buf.len) {
+                    @memcpy(fn_name_buf[0..name.base.len], name.base);
+                    fn_name_buf[name.base.len] = 0;
+                    const fn_name_z = fn_name_buf[0..name.base.len :0];
+                    if (c.LLVMGetNamedFunction(self.module, fn_name_z)) |fn_val| {
+                        break :blk fn_val;
+                    }
+                }
+                std.debug.print("UnsupportedGrinVal: Var {s}_{d} not found in params, tag table, or module\n", .{ name.base, name.unique.value });
                 return error.UnsupportedGrinVal;
             },
             .ConstTagNode => |ctn| blk: {
@@ -755,8 +812,7 @@ pub const GrinTranslator = struct {
                     // Fields are stored as opaque ptr; box integers by alloca+store.
                     const field_ty = c.LLVMTypeOf(raw_val);
                     const kind = c.LLVMGetTypeKind(field_ty);
-                    // LLVMPointerTypeKind == 7 in LLVM 19 C API (c_int constant, not enum)
-                    const is_ptr = kind == 7;
+                    const is_ptr = kind == c.LLVMPointerTypeKind;
                     if (!is_ptr) {
                         // Box the scalar: alloca + store, use the pointer.
                         const slot = c.LLVMBuildAlloca(self.builder, field_ty, "box");
@@ -803,8 +859,7 @@ pub const GrinTranslator = struct {
         const scrut_llvm = try self.translateValToLlvm(scrutinee);
         const scrut_ty = c.LLVMTypeOf(scrut_llvm);
         const scrut_kind = c.LLVMGetTypeKind(scrut_ty);
-        // LLVMPointerTypeKind == 7 in LLVM 19 C API (c_int constant, not enum)
-        const is_ptr = scrut_kind == 7;
+        const is_ptr = scrut_kind == c.LLVMPointerTypeKind;
 
         const tag_val: llvm.Value = if (is_ptr) blk: {
             // Load the first field (i64 tag) from the heap node.
@@ -911,7 +966,7 @@ pub const GrinTranslator = struct {
                         gep,
                         nullTerminate(field_name.base).ptr,
                     );
-                    try self.params.put(field_name.base, loaded);
+                    try self.params.put(field_name.unique.value, loaded);
                 }
             }
 

--- a/src/core/desugar.zig
+++ b/src/core/desugar.zig
@@ -5,6 +5,7 @@ const infer_mod = @import("../typechecker/infer.zig");
 const htype_mod = @import("../typechecker/htype.zig");
 const env_mod = @import("../typechecker/env.zig");
 const Known = @import("../naming/known.zig");
+const naming_mod = @import("../naming/unique.zig");
 const Name = ast_mod.Name;
 const SourceSpan = ast_mod.SourceSpan;
 const SourcePos = ast_mod.SourcePos;
@@ -14,6 +15,11 @@ pub const DesugarCtx = struct {
     alloc: std.mem.Allocator,
     types: *const infer_mod.ModuleTypes,
     diags: *DiagnosticCollector,
+    /// Used to generate fresh unique values for synthetic binders (lambda
+    /// parameters created by the pattern-match compiler, constructor field
+    /// binders, etc.). Each synthetic binder must get a globally unique ID so
+    /// that the GRIN translator can distinguish them in its `var_map`.
+    u_supply: *naming_mod.UniqueSupply,
 };
 
 // Synthetic span for desugared constructs
@@ -48,11 +54,13 @@ pub fn desugarModule(
     module: renamer_mod.RenamedModule,
     module_types: *const infer_mod.ModuleTypes,
     diags: *DiagnosticCollector,
+    u_supply: *naming_mod.UniqueSupply,
 ) std.mem.Allocator.Error!ast_mod.CoreProgram {
     var ctx = DesugarCtx{
         .alloc = alloc,
         .types = module_types,
         .diags = diags,
+        .u_supply = u_supply,
     };
 
     var binds = std.ArrayListUnmanaged(ast_mod.Bind){};
@@ -638,11 +646,14 @@ fn desugarMatch(
     const arg_tys = try getArgTypes(ctx.alloc, core_ty, num_args);
 
     // Create fresh lambda binders: `arg_0, arg_1, ...`
+    // Each binder gets a fresh unique so that the GRIN translator's var_map
+    // can distinguish parameters that share the same base name (e.g. when the
+    // pattern-match compiler generates multiple `arg_N` parameters).
     var binders = try ctx.alloc.alloc(ast_mod.Id, num_args);
     for (0..num_args) |i| {
         const arg_name = try std.fmt.allocPrint(ctx.alloc, "arg_{d}", .{i});
         binders[i] = .{
-            .name = .{ .base = arg_name, .unique = .{ .value = 0 } },
+            .name = .{ .base = arg_name, .unique = ctx.u_supply.fresh() },
             .ty = arg_tys[i],
             .span = syntheticSpan(),
         };
@@ -797,7 +808,7 @@ fn applyPat(
                 var sub_pat = c.args[field_i];
                 while (sub_pat == .Paren) sub_pat = sub_pat.Paren.*;
 
-                const field_binder = try freshFieldBinder(ctx.alloc, arg_idx, depth, field_i);
+                const field_binder = try freshFieldBinder(ctx.alloc, ctx.u_supply, arg_idx, depth, field_i);
 
                 // The binder always captures the field value, regardless of whether
                 // the sub-pattern matches.
@@ -891,7 +902,7 @@ fn applyPat(
                 var sub_pat = elems[field_i];
                 while (sub_pat == .Paren) sub_pat = sub_pat.Paren.*;
 
-                const field_binder = try freshFieldBinder(ctx.alloc, arg_idx, depth, field_i);
+                const field_binder = try freshFieldBinder(ctx.alloc, ctx.u_supply, arg_idx, depth, field_i);
                 con_binders[field_i] = field_binder;
 
                 switch (sub_pat) {
@@ -1009,16 +1020,20 @@ fn applyPat(
 }
 
 /// Allocate a fresh `Id` for a constructor field binder.
-/// Name format: `_field_{arg_idx}_{depth}_{field_i}` — unique within a match clause.
+/// Create a fresh binder for a constructor field in a pattern match.
+///
+/// Each binder gets a fresh unique from `u_supply` so that the downstream
+/// passes can distinguish field binders that share a base name pattern.
 fn freshFieldBinder(
     alloc: std.mem.Allocator,
+    u_supply: *naming_mod.UniqueSupply,
     arg_idx: usize,
     depth: usize,
     field_i: usize,
 ) !ast_mod.Id {
     const name = try std.fmt.allocPrint(alloc, "_field_{d}_{d}_{d}", .{ arg_idx, depth, field_i });
     return .{
-        .name = .{ .base = name, .unique = .{ .value = 0 } },
+        .name = .{ .base = name, .unique = u_supply.fresh() },
         .ty = dummyCoreType(),
         .span = syntheticSpan(),
     };
@@ -1060,10 +1075,12 @@ test "desugarMatch: Tier 1 literal and wildcard" {
     var diags = DiagnosticCollector.init();
     defer diags.deinit(alloc);
 
+    var u_supply = naming_mod.UniqueSupply{};
     var ctx = DesugarCtx{
         .alloc = alloc,
         .types = &types,
         .diags = &diags,
+        .u_supply = &u_supply,
     };
 
     // f 0 = "zero"
@@ -1141,10 +1158,12 @@ test "desugarExpr: Var and Lit" {
     var diags = DiagnosticCollector.init();
     defer diags.deinit(alloc);
 
+    var u_supply = naming_mod.UniqueSupply{};
     var ctx = DesugarCtx{
         .alloc = alloc,
         .types = &types,
         .diags = &diags,
+        .u_supply = &u_supply,
     };
 
     // Add a local binder type mapping
@@ -1223,10 +1242,12 @@ test "desugarExpr: List desugars to Cons/Nil chain" {
         try types.schemes.put(alloc, Known.Con.Nil.unique, .{ .binders = binders, .constraints = &.{}, .body = nil_body });
     }
 
+    var u_supply = naming_mod.UniqueSupply{};
     var ctx = DesugarCtx{
         .alloc = alloc,
         .types = &types,
         .diags = &diags,
+        .u_supply = &u_supply,
     };
 
     // Test empty list: [] desugars to Var(Nil)
@@ -1297,10 +1318,12 @@ test "desugarExpr: Paren unwraps inner expression" {
     const x_name = testName("x", 42);
     try types.local_binders.put(alloc, x_name.unique, int_h_ty);
 
+    var u_supply = naming_mod.UniqueSupply{};
     var ctx = DesugarCtx{
         .alloc = alloc,
         .types = &types,
         .diags = &diags,
+        .u_supply = &u_supply,
     };
 
     const inner = try alloc.create(renamer_mod.RExpr);
@@ -1316,8 +1339,8 @@ test "desugarExpr: Paren unwraps inner expression" {
 // ── Tier 2 pattern match tests ──────────────────────────────────────────────
 
 /// Helper: build a DesugarCtx backed by an ArenaAllocator for pattern tests.
-fn makeCtx(alloc: std.mem.Allocator, types: *infer_mod.ModuleTypes, diags: *DiagnosticCollector) DesugarCtx {
-    return .{ .alloc = alloc, .types = types, .diags = diags };
+fn makeCtx(alloc: std.mem.Allocator, types: *infer_mod.ModuleTypes, diags: *DiagnosticCollector, u_supply: *naming_mod.UniqueSupply) DesugarCtx {
+    return .{ .alloc = alloc, .types = types, .diags = diags, .u_supply = u_supply };
 }
 
 /// Helper: build a simple `Con_name -> T` function type for single-arg functions.
@@ -1341,7 +1364,8 @@ test "desugarMatch: Tier 2 constructor pattern with Var sub-pattern" {
     defer types.deinit(alloc);
     var diags = DiagnosticCollector.init();
     defer diags.deinit(alloc);
-    var ctx = makeCtx(alloc, &types, &diags);
+    var u_supply = naming_mod.UniqueSupply{};
+    var ctx = makeCtx(alloc, &types, &diags, &u_supply);
 
     const maybe_ty_name = testName("Maybe", 10);
     const int_ty_name = testName("Int", 1);
@@ -1407,7 +1431,8 @@ test "desugarMatch: Tier 2 as-pattern" {
     defer types.deinit(alloc);
     var diags = DiagnosticCollector.init();
     defer diags.deinit(alloc);
-    var ctx = makeCtx(alloc, &types, &diags);
+    var u_supply = naming_mod.UniqueSupply{};
+    var ctx = makeCtx(alloc, &types, &diags, &u_supply);
 
     const maybe_ty_name = testName("Maybe", 10);
     const int_ty_name = testName("Int", 1);
@@ -1471,7 +1496,8 @@ test "desugarMatch: Tier 2 tuple pattern" {
     defer types.deinit(alloc);
     var diags = DiagnosticCollector.init();
     defer diags.deinit(alloc);
-    var ctx = makeCtx(alloc, &types, &diags);
+    var u_supply = naming_mod.UniqueSupply{};
+    var ctx = makeCtx(alloc, &types, &diags, &u_supply);
 
     const int_ty_name = testName("Int", 1);
     const int_ty = ast_mod.CoreType{ .TyCon = .{ .name = int_ty_name, .args = &.{} } };

--- a/src/grin/translate.zig
+++ b/src/grin/translate.zig
@@ -55,6 +55,11 @@ const TranslateCtx = struct {
     // Maps function names to their arity (number of parameters).
     // Used for partial/over-application handling (issue #317).
     arity_map: std.AutoHashMapUnmanaged(u64, u32),
+    // Maps data constructor uniques to their field count (number of value
+    // arguments, excluding type arguments).  Built from CoreProgram.data_decls.
+    // Used to distinguish constructor applications from function calls in
+    // translateApp, and to emit ValTag for nullary constructors in translateExpr.
+    con_map: std.AutoHashMapUnmanaged(u64, u32),
     // Counter for generating fresh GRIN variable names.
     name_counter: u64 = 0,
 
@@ -63,12 +68,24 @@ const TranslateCtx = struct {
             .alloc = alloc,
             .var_map = .{},
             .arity_map = .{},
+            .con_map = .{},
         };
     }
 
     pub fn deinit(self: *TranslateCtx) void {
         self.var_map.deinit(self.alloc);
         self.arity_map.deinit(self.alloc);
+        self.con_map.deinit(self.alloc);
+    }
+
+    /// Return true if the given unique belongs to a known data constructor.
+    fn isConstructor(self: TranslateCtx, unique: u64) bool {
+        return self.con_map.contains(unique);
+    }
+
+    /// Return the field count for a constructor, or null if unknown.
+    fn conFieldCount(self: TranslateCtx, unique: u64) ?u32 {
+        return self.con_map.get(unique);
     }
 
     /// Generate a fresh GRIN variable name.
@@ -147,6 +164,55 @@ fn buildArityMap(alloc: std.mem.Allocator, core_prog: CoreProgram) !std.AutoHash
     return arity_map;
 }
 
+/// Build a map from data constructor uniques to their field counts.
+///
+/// The field count of a constructor is the number of value-level arguments
+/// its type has, i.e. the number of leading `FunTy` arrows after stripping
+/// `ForAllTy` quantifiers.
+///
+/// Example:
+///   Nil  :: forall a. List a         → 0 fields
+///   Cons :: forall a. a -> List a -> List a  → 2 fields
+fn buildConMap(alloc: std.mem.Allocator, core_prog: CoreProgram) !std.AutoHashMapUnmanaged(u64, u32) {
+    var con_map = std.AutoHashMapUnmanaged(u64, u32){};
+    errdefer con_map.deinit(alloc);
+
+    for (core_prog.data_decls) |decl| {
+        for (decl.constructors) |con| {
+            const unique = con.name.unique.value;
+            const n_fields = countConFields(con.ty);
+            try con_map.put(alloc, unique, n_fields);
+        }
+    }
+
+    return con_map;
+}
+
+/// Count the value-level field count of a constructor type by counting
+/// leading FunTy arrows, stripping ForAllTy quantifiers first.
+fn countConFields(ty: CoreType) u32 {
+    var current = ty;
+    // Strip leading ForAllTy quantifiers.
+    while (true) {
+        switch (current) {
+            .ForAllTy => |fa| current = fa.body.*,
+            else => break,
+        }
+    }
+    // Count FunTy arrows.
+    var fields: u32 = 0;
+    while (true) {
+        switch (current) {
+            .FunTy => |ft| {
+                fields += 1;
+                current = ft.res.*;
+            },
+            else => break,
+        }
+    }
+    return fields;
+}
+
 /// Count the number of lambda parameters in a Core expression.
 fn countLambdaArity(expr: *const CoreExpr) !u32 {
     var arity: u32 = 0;
@@ -169,13 +235,23 @@ pub fn translateProgram(alloc: std.mem.Allocator, core_prog: CoreProgram) !GrinP
     var arity_map = try buildArityMap(alloc, core_prog);
     defer arity_map.deinit(alloc);
 
+    // Build the constructor map from data declarations.
+    var con_map = try buildConMap(alloc, core_prog);
+    defer con_map.deinit(alloc);
+
     var ctx = TranslateCtx.init(alloc);
     defer ctx.deinit();
 
-    // Copy the arity map into the context
+    // Copy the arity map into the context.
     var iter = arity_map.iterator();
     while (iter.next()) |entry| {
         try ctx.arity_map.put(alloc, entry.key_ptr.*, entry.value_ptr.*);
+    }
+
+    // Copy the constructor map into the context.
+    var con_iter = con_map.iterator();
+    while (con_iter.next()) |entry| {
+        try ctx.con_map.put(alloc, entry.key_ptr.*, entry.value_ptr.*);
     }
 
     var defs = std.ArrayListUnmanaged(GrinDef){};
@@ -249,9 +325,26 @@ fn translateExpr(ctx: *TranslateCtx, expr: *const CoreExpr) !*GrinExpr {
 
     switch (expr.*) {
         .Var => |v| {
-            // Variable reference becomes a variable value.
-            const grin_name = try ctx.getCoreVar(v.name);
-            new_expr.* = .{ .Return = .{ .Var = grin_name } };
+            const unique = v.name.unique.value;
+            if (ctx.conFieldCount(unique)) |n_fields| {
+                if (n_fields == 0) {
+                    // Nullary constructor (e.g. Nil, True, Nothing): emit a bare tag value.
+                    // In GRIN, nullary constructors are immediate tag constants — no heap
+                    // allocation needed.  The LLVM backend emits an i64 discriminant for them.
+                    const tag = GrinTag{ .tag_type = .Con, .name = v.name };
+                    new_expr.* = .{ .Return = .{ .ValTag = tag } };
+                } else {
+                    // Non-nullary constructor used as a value (e.g. partially applied).
+                    // Fall back to a variable reference; proper closure support is tracked in:
+                    // https://github.com/adinapoli/rusholme/issues/386
+                    const grin_name = try ctx.getCoreVar(v.name);
+                    new_expr.* = .{ .Return = .{ .Var = grin_name } };
+                }
+            } else {
+                // Ordinary variable reference.
+                const grin_name = try ctx.getCoreVar(v.name);
+                new_expr.* = .{ .Return = .{ .Var = grin_name } };
+            }
         },
         .Lit => |l| {
             // Literal becomes a return with literal value.
@@ -358,6 +451,14 @@ fn translateApp(ctx: *TranslateCtx, app_expr: *const CoreExpr) anyerror!*GrinExp
         }
     }
 
+    // Check whether the function head is a data constructor *before* translating
+    // it.  We need the original Core unique to look up the con_map, because
+    // translateExpr assigns a fresh GRIN unique that won't match.
+    const core_con_unique: ?u64 = switch (fn_expr.*) {
+        .Var => |v| if (ctx.isConstructor(v.name.unique.value)) v.name.unique.value else null,
+        else => null,
+    };
+
     // Translate the function head.
     const translated_fn = try translateExpr(ctx, fn_expr);
 
@@ -416,6 +517,47 @@ fn translateApp(ctx: *TranslateCtx, app_expr: *const CoreExpr) anyerror!*GrinExp
                     .complex_expr = arg_result,
                 });
             },
+        }
+    }
+
+    // ── Constructor application ────────────────────────────────────────────
+    //
+    // If the function head is a known data constructor, emit a heap allocation
+    // instead of a function call:
+    //   `Cons x xs`  →  store (CCons [x, xs])
+    //
+    // This handles saturated constructor applications.  Partial application of
+    // constructors (tracked in https://github.com/adinapoli/rusholme/issues/386)
+    // falls through to the general partial-application path below.
+    if (core_con_unique) |con_unique| {
+        if (fn_name_and_arity.name) |fn_name| {
+            const n_fields = ctx.conFieldCount(con_unique).?;
+            const arg_count = @as(u32, @intCast(args.items.len));
+            if (arg_count == n_fields) {
+                // Saturated constructor application → Store (ConstTagNode).
+                // Use the original Core name so the tag carries the stable unique.
+                const core_name = switch (fn_expr.*) {
+                    .Var => |v| v.name,
+                    else => fn_name,
+                };
+                const tag = GrinTag{ .tag_type = .Con, .name = core_name };
+                const store_expr = try ctx.alloc.create(GrinExpr);
+                store_expr.* = .{ .Store = .{ .ConstTagNode = .{
+                    .tag = tag,
+                    .fields = grin_args,
+                } } };
+                const ptr_var = try ctx.freshName("node");
+                const return_expr = try ctx.alloc.create(GrinExpr);
+                return_expr.* = .{ .Return = .{ .Var = ptr_var } };
+                const bind_expr = try ctx.alloc.create(GrinExpr);
+                bind_expr.* = .{ .Bind = .{
+                    .lhs = store_expr,
+                    .pat = .{ .Var = ptr_var },
+                    .rhs = return_expr,
+                } };
+                return try wrapWithPendingBinds(ctx, bind_expr, pending_binds.items);
+            }
+            // Under-saturated constructor: fall through to partial-application path.
         }
     }
 
@@ -589,17 +731,18 @@ fn translateLet(ctx: *TranslateCtx, let_expr: *const CoreLet) anyerror!*GrinExpr
                 break :blk e;
             };
 
-            // Bind the stored value to a variable.
-            const fresh_ptr = try ctx.freshName("p");
+            // Bind the stored value using the actual let-bound binder name so the
+            // body's references (via var_map) resolve correctly.
+            const binder_name = try ctx.mapBinder(&pair.binder);
 
             // Translate the body.
             const body_expr = try translateExpr(ctx, let_expr.body);
 
-            // Create the bind expression: store X >>= \p -> body
+            // Create the bind expression: store X >>= \binder_name -> body
             const bind_expr = try ctx.alloc.create(GrinExpr);
             bind_expr.* = .{ .Bind = .{
                 .lhs = store_expr,
-                .pat = .{ .Var = fresh_ptr },
+                .pat = .{ .Var = binder_name },
                 .rhs = body_expr,
             } };
 

--- a/src/main.zig
+++ b/src/main.zig
@@ -379,7 +379,7 @@ fn cmdCore(allocator: std.mem.Allocator, io: Io, file_path: []const u8) !void {
     }
 
     // ── Desugar ────────────────────────────────────────────────────────
-    const core_prog = try rusholme.core.desugar.desugarModule(arena_alloc, renamed, &module_types, &diags);
+    const core_prog = try rusholme.core.desugar.desugarModule(arena_alloc, renamed, &module_types, &diags, &u_supply);
     if (diags.hasErrors()) {
         try renderDiagnostics(allocator, io, &diags, file_id, file_path, source);
         std.process.exit(1);
@@ -464,7 +464,7 @@ fn cmdGrin(allocator: std.mem.Allocator, io: Io, file_path: []const u8) !void {
     }
 
     // ── Desugar ────────────────────────────────────────────────────────
-    const core_prog = try rusholme.core.desugar.desugarModule(arena_alloc, renamed, &module_types, &diags);
+    const core_prog = try rusholme.core.desugar.desugarModule(arena_alloc, renamed, &module_types, &diags, &u_supply);
     if (diags.hasErrors()) {
         try renderDiagnostics(allocator, io, &diags, file_id, file_path, source);
         std.process.exit(1);
@@ -563,7 +563,7 @@ fn cmdLl(allocator: std.mem.Allocator, io: Io, file_path: []const u8) !void {
     }
 
     // ── Desugar ────────────────────────────────────────────────────────
-    const core_prog = try rusholme.core.desugar.desugarModule(arena_alloc, renamed, &module_types, &diags);
+    const core_prog = try rusholme.core.desugar.desugarModule(arena_alloc, renamed, &module_types, &diags, &u_supply);
     if (diags.hasErrors()) {
         try renderDiagnostics(allocator, io, &diags, file_id, file_path, source);
         std.process.exit(1);
@@ -675,7 +675,7 @@ fn cmdBuild(allocator: std.mem.Allocator, io: Io, file_path: []const u8, output_
     }
 
     // ── Desugar ────────────────────────────────────────────────────────
-    const core_prog = try rusholme.core.desugar.desugarModule(arena_alloc, renamed, &module_types, &diags);
+    const core_prog = try rusholme.core.desugar.desugarModule(arena_alloc, renamed, &module_types, &diags, &u_supply);
     if (diags.hasErrors()) {
         try renderDiagnostics(allocator, io, &diags, file_id, file_path, source);
         std.process.exit(1);

--- a/tests/golden/ghc_016_tier1_patterns.core.golden
+++ b/tests/golden/ghc_016_tier1_patterns.core.golden
@@ -1,14 +1,14 @@
 === Core Program (1 data, 2 bindings) ===
 data Color_1005 = (Red_1000 :: Color_1005) | (Green_1001 :: Color_1005) | (Blue_1002 :: Color_1005)
 
-(isRed_1003 :: Color_1005 -> Bool_104) = \(arg_0 :: Color_1005) -> case arg_0 of (arg_0 :: Color_1005) {
+(isRed_1003 :: Color_1005 -> Bool_104) = \(arg_0_1006 :: Color_1005) -> case arg_0_1006 of (arg_0_1006 :: Color_1005) {
   Red_1000 -> True_200
   _ -> False_201
 }
 
-(describeNumber_1004 :: Int_100 -> []_110 Char_105) = \(arg_0 :: Int_100) -> case arg_0 of (arg_0 :: Int_100) {
+(describeNumber_1004 :: Int_100 -> []_110 Char_105) = \(arg_0_1007 :: Int_100) -> case arg_0_1007 of (arg_0_1007 :: Int_100) {
   0 -> "zero"
-  _ -> case arg_0 of (arg_0 :: Int_100) {
+  _ -> case arg_0_1007 of (arg_0_1007 :: Int_100) {
     1 -> "one"
     _ -> "many"
   }

--- a/tests/golden/ghc_017_tier2_patterns.core.golden
+++ b/tests/golden/ghc_017_tier2_patterns.core.golden
@@ -3,29 +3,29 @@ data Tree_1007 a = (Leaf_1000 :: forall a_1013. Tree_1007 a_1013) | (Node_1001 :
 
 data Wrapper_1008 a = (Wrap_1002 :: forall a_1014. a_1014 -> Wrapper_1008 a_1014)
 
-(unwrapNode_1003 :: (Tree_1007 Int_100) -> Int_100) = \(arg_0 :: Tree_1007 Int_100) -> case arg_0 of (arg_0 :: Tree_1007 Int_100) {
-  Node_1001 (_field_0_0_0 :: _t) (x_1009 :: Int_100) (_field_0_0_2 :: _t) -> x_1009
-  _ -> case arg_0 of (arg_0 :: Tree_1007 Int_100) {
+(unwrapNode_1003 :: (Tree_1007 Int_100) -> Int_100) = \(arg_0_1015 :: Tree_1007 Int_100) -> case arg_0_1015 of (arg_0_1015 :: Tree_1007 Int_100) {
+  Node_1001 (_field_0_0_0_1018 :: _t) (x_1009 :: Int_100) (_field_0_0_2_1016 :: _t) -> x_1009
+  _ -> case arg_0_1015 of (arg_0_1015 :: Tree_1007 Int_100) {
     Leaf_1000 -> 0
     _ -> "Non-exhaustive patterns in function"
   }
 }
 
-(leftValue_1004 :: (Tree_1007 Int_100) -> Int_100) = \(arg_0 :: Tree_1007 Int_100) -> case arg_0 of (arg_0 :: Tree_1007 Int_100) {
-  Node_1001 (_field_0_0_0 :: _t) (_field_0_0_1 :: _t) (_field_0_0_2 :: _t) -> case _field_0_0_0 of (_field_0_0_0 :: _t) {
-    Node_1001 (_field_0_1_0 :: _t) (x_1010 :: Int_100) (_field_0_1_2 :: _t) -> x_1010
+(leftValue_1004 :: (Tree_1007 Int_100) -> Int_100) = \(arg_0_1019 :: Tree_1007 Int_100) -> case arg_0_1019 of (arg_0_1019 :: Tree_1007 Int_100) {
+  Node_1001 (_field_0_0_0_1022 :: _t) (_field_0_0_1_1021 :: _t) (_field_0_0_2_1020 :: _t) -> case _field_0_0_0_1022 of (_field_0_0_0_1022 :: _t) {
+    Node_1001 (_field_0_1_0_1025 :: _t) (x_1010 :: Int_100) (_field_0_1_2_1023 :: _t) -> x_1010
     _ -> 0
   }
   _ -> 0
 }
 
-(wrapOrDefault_1005 :: (Wrapper_1008 Int_100) -> Wrapper_1008 Int_100) = \(arg_0 :: Wrapper_1008 Int_100) -> let (w_1011 :: Wrapper_1008 Int_100) = arg_0
-in case arg_0 of (arg_0 :: Wrapper_1008 Int_100) {
-  Wrap_1002 (_field_0_0_0 :: _t) -> w_1011
+(wrapOrDefault_1005 :: (Wrapper_1008 Int_100) -> Wrapper_1008 Int_100) = \(arg_0_1026 :: Wrapper_1008 Int_100) -> let (w_1011 :: Wrapper_1008 Int_100) = arg_0_1026
+in case arg_0_1026 of (arg_0_1026 :: Wrapper_1008 Int_100) {
+  Wrap_1002 (_field_0_0_0_1027 :: _t) -> w_1011
   _ -> Wrap_1002 0
 }
 
-(sumPair_1006 :: ((,)_209 Int_100 Int_100) -> Int_100) = \(arg_0 :: (,)_209 Int_100 Int_100) -> case arg_0 of (arg_0 :: (,)_209 Int_100 Int_100) {
-  (,) (a_1012 :: Int_100) (_field_0_0_1 :: _t) -> a_1012
+(sumPair_1006 :: ((,)_209 Int_100 Int_100) -> Int_100) = \(arg_0_1028 :: (,)_209 Int_100 Int_100) -> case arg_0_1028 of (arg_0_1028 :: (,)_209 Int_100 Int_100) {
+  (,) (a_1012 :: Int_100) (_field_0_0_1_1029 :: _t) -> a_1012
   _ -> "Non-exhaustive patterns in function"
 }

--- a/tests/golden_test_runner.zig
+++ b/tests/golden_test_runner.zig
@@ -151,7 +151,7 @@ fn pipelineToCore(allocator: std.mem.Allocator, source: []const u8) ![]const u8 
     if (diags.hasErrors()) return error.TypecheckError;
 
     // ── Desugar ──────────────────────────────────────────────────────────
-    const core_prog = try rusholme.core.desugar.desugarModule(arena_alloc, renamed, &module_types, &diags);
+    const core_prog = try rusholme.core.desugar.desugarModule(arena_alloc, renamed, &module_types, &diags, &u_supply);
     if (diags.hasErrors()) return error.DesugarError;
 
     // ── Pretty-print Core to a heap string ───────────────────────────────

